File: sdk/node/src/storage/sqlite.ts
import Database from 'better-sqlite3';
import type { StorageAdapter, Agent, EnrollmentToken, Challenge } from '../types.js';

export class SQLiteStorageAdapter implements StorageAdapter {
  private db: Database.Database;

  agents: StorageAdapter['agents'];
  enrollmentTokens: StorageAdapter['enrollmentTokens'];
  challenges: StorageAdapter['challenges'];

  constructor(path: string) {
    this.db = new Database(path);
    this.db.pragma('journal_mode = WAL');
    this.initTables();

    this.agents = {
      create: this.createAgent.bind(this),
      get: this.getAgent.bind(this),
      updateStatus: this.updateAgentStatus.bind(this),
      listByUser: this.listAgentsByUser.bind(this),
    };

    this.enrollmentTokens = {
      create: this.createEnrollmentToken.bind(this),
      validate: this.validateEnrollmentToken.bind(this),
      burn: this.burnEnrollmentToken.bind(this),
    };

    this.challenges = {
      create: this.createChallenge.bind(this),
      validate: this.validateChallenge.bind(this),
      burn: this.burnChallenge.bind(this),
    };
  }

  private initTables(): void {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS sigil_agents (
        id TEXT PRIMARY KEY,
        name TEXT NOT NULL,
        external_user_id TEXT NOT NULL,
        public_key BLOB,
        key_fingerprint TEXT,
        platform TEXT,
        status TEXT NOT NULL DEFAULT 'pending_enrollment',
        enrolled_at TEXT,
        last_auth_at TEXT,
        key_expires_at TEXT,
        created_at TEXT NOT NULL
      );

      CREATE TABLE IF NOT EXISTS sigil_enrollment_tokens (
        token_hash TEXT PRIMARY KEY,
        agent_id TEXT NOT NULL,
        expires_at TEXT NOT NULL,
        used INTEGER NOT NULL DEFAULT 0
      );

      CREATE TABLE IF NOT EXISTS sigil_challenges (
        challenge TEXT PRIMARY KEY,
        agent_id TEXT NOT NULL,
        expires_at TEXT NOT NULL,
        used INTEGER NOT NULL DEFAULT 0
      );
    `);
  }

  // ─── Agents ──────────────────────────────────────────────────────────

  private async createAgent(input: Omit<Agent, 'createdAt'>): Promise<Agent> {
    const createdAt = new Date();
    const stmt = this.db.prepare(`
      INSERT INTO sigil_agents (id, name, external_user_id, public_key, key_fingerprint, platform, status, enrolled_at, last_auth_at, key_expires_at, created_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);
    stmt.run(
      input.id,
      input.name,
      input.externalUserId,
      input.publicKey ?? null,
      input.keyFingerprint ?? null,
      input.platform ?? null,
      input.status,
      input.enrolledAt ? input.enrolledAt.toISOString() : null,
      input.lastAuthAt ? input.lastAuthAt.toISOString() : null,
      input.keyExpiresAt ? input.keyExpiresAt.toISOString() : null,
      createdAt.toISOString(),
    );
    return { ...input, createdAt };
  }

  private async getAgent(id: string): Promise<Agent | null> {
    const row = this.db.prepare('SELECT * FROM sigil_agents WHERE id = ?').get(id) as
      | Record<string, unknown>
      | undefined;
    if (!row) return null;
    return this.rowToAgent(row);
  }

  private async updateAgentStatus(
    id: string,
    status: Agent['status'],
    fields?: Partial<Agent>,
  ): Promise<void> {
    const sets: string[] = ['status = ?'];
    const values: unknown[] = [status];

    if (fields) {
      if (fields.publicKey !== undefined) {
        sets.push('public_key = ?');
        values.push(fields.publicKey ?? null);
      }
      if (fields.keyFingerprint !== undefined) {
        sets.push('key_fingerprint = ?');
        values.push(fields.keyFingerprint ?? null);
      }
      if (fields.platform !== undefined) {
        sets.push('platform = ?');
        values.push(fields.platform ?? null);
      }
      if (fields.enrolledAt !== undefined) {
        sets.push('enrolled_at = ?');
        values.push(fields.enrolledAt ? fields.enrolledAt.toISOString() : null);
      }
      if (fields.lastAuthAt !== undefined) {
        sets.push('last_auth_at = ?');
        values.push(fields.lastAuthAt ? fields.lastAuthAt.toISOString() : null);
      }
      if (fields.keyExpiresAt !== undefined) {
        sets.push('key_expires_at = ?');
        values.push(fields.keyExpiresAt ? fields.keyExpiresAt.toISOString() : null);
      }
      if (fields.name !== undefined) {
        sets.push('name = ?');
        values.push(fields.name);
      }
    }

    values.push(id);
    this.db.prepare(`UPDATE sigil_agents SET ${sets.join(', ')} WHERE id = ?`).run(...values);
  }

  private async listAgentsByUser(userId: string): Promise<Agent[]> {
    const rows = this.db
      .prepare('SELECT * FROM sigil_agents WHERE external_user_id = ?')
      .all(userId) as Record<string, unknown>[];
    return rows.map((row) => this.rowToAgent(row));
  }

  private rowToAgent(row: Record<string, unknown>): Agent {
    return {
      id: row.id as string,
      name: row.name as string,
      externalUserId: row.external_user_id as string,
      publicKey: row.public_key ? Buffer.from(row.public_key as Buffer) : null,
      keyFingerprint: (row.key_fingerprint as string) ?? null,
      platform: (row.platform as string) ?? null,
      status: row.status as Agent['status'],
      enrolledAt: row.enrolled_at ? new Date(row.enrolled_at as string) : null,
      lastAuthAt: row.last_auth_at ? new Date(row.last_auth_at as string) : null,
      keyExpiresAt: row.key_expires_at ? new Date(row.key_expires_at as string) : null,
      createdAt: new Date(row.created_at as string),
    };
  }

  // ─── Enrollment Tokens ──────────────────────────────────────────────

  private async createEnrollmentToken(token: EnrollmentToken): Promise<void> {
    this.db
      .prepare(
        `INSERT INTO sigil_enrollment_tokens (token_hash, agent_id, expires_at, used)
       VALUES (?, ?, ?, ?)`,
      )
      .run(token.tokenHash, token.agentId, token.expiresAt.toISOString(), token.used ? 1 : 0);
  }

  private async validateEnrollmentToken(tokenHash: string): Promise<EnrollmentToken | null> {
    const row = this.db
      .prepare(
        `SELECT * FROM sigil_enrollment_tokens
       WHERE token_hash = ? AND used = 0 AND expires_at > ?`,
      )
      .get(tokenHash, new Date().toISOString()) as Record<string, unknown> | undefined;
    if (!row) return null;
    return this.rowToEnrollmentToken(row);
  }

  private async burnEnrollmentToken(tokenHash: string): Promise<void> {
    this.db
      .prepare('UPDATE sigil_enrollment_tokens SET used = 1 WHERE token_hash = ?')
      .run(tokenHash);
  }

  private rowToEnrollmentToken(row: Record<string, unknown>): EnrollmentToken {
    return {
      tokenHash: row.token_hash as string,
      agentId: row.agent_id as string,
      expiresAt: new Date(row.expires_at as string),
      used: (row.used as number) === 1,
    };
  }

  // ─── Challenges ─────────────────────────────────────────────────────

  private async createChallenge(challenge: Challenge): Promise<void> {
    this.db
      .prepare(
        `INSERT INTO sigil_challenges (challenge, agent_id, expires_at, used)
       VALUES (?, ?, ?, ?)`,
      )
      .run(
        challenge.challenge,
        challenge.agentId,
        challenge.expiresAt.toISOString(),
        challenge.used ? 1 : 0,
      );
  }

  private async validateChallenge(challenge: string): Promise<Challenge | null> {
    const row = this.db
      .prepare(
        `SELECT * FROM sigil_challenges
       WHERE challenge = ? AND used = 0 AND expires_at > ?`,
      )
      .get(challenge, new Date().toISOString()) as Record<string, unknown> | undefined;
    if (!row) return null;
    return this.rowToChallenge(row);
  }

  private async burnChallenge(challenge: string): Promise<void> {
    this.db
      .prepare('UPDATE sigil_challenges SET used = 1 WHERE challenge = ?')
      .run(challenge);
  }

  private rowToChallenge(row: Record<string, unknown>): Challenge {
    return {
      challenge: row.challenge as string,
      agentId: row.agent_id as string,
      expiresAt: new Date(row.expires_at as string),
      used: (row.used as number) === 1,
    };
  }

  // ─── Cleanup ────────────────────────────────────────────────────────

  async cleanup(): Promise<void> {
    const now = new Date().toISOString();
    this.db.prepare('DELETE FROM sigil_challenges WHERE expires_at <= ? OR used = 1').run(now);
    this.db
      .prepare('DELETE FROM sigil_enrollment_tokens WHERE expires_at <= ? OR used = 1')
      .run(now);
  }

  /** Close the database connection. */
  close(): void {
    this.db.close();
  }
}

File: sdk/python/sigil/storage/sqlite.py
"""SQLite storage adapter using aiosqlite for async access."""

from __future__ import annotations

from datetime import datetime

import aiosqlite

from sigil.types import Agent, Challenge, EnrollmentToken

_CREATE_TABLES = """
CREATE TABLE IF NOT EXISTS sigil_agents (
    id TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    external_user_id TEXT NOT NULL,
    public_key BLOB,
    key_fingerprint TEXT,
    platform TEXT,
    status TEXT NOT NULL DEFAULT 'pending_enrollment',
    enrolled_at TEXT,
    last_auth_at TEXT,
    key_expires_at TEXT,
    created_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS sigil_enrollment_tokens (
    token_hash TEXT PRIMARY KEY,
    agent_id TEXT NOT NULL,
    expires_at TEXT NOT NULL,
    used INTEGER NOT NULL DEFAULT 0
);

CREATE TABLE IF NOT EXISTS sigil_challenges (
    challenge TEXT PRIMARY KEY,
    agent_id TEXT NOT NULL,
    expires_at TEXT NOT NULL,
    used INTEGER NOT NULL DEFAULT 0
);
"""


def _dt_to_str(dt: datetime | None) -> str | None:
    """Convert a datetime to ISO format string, or None."""
    return dt.isoformat() if dt else None


def _str_to_dt(s: str | None) -> datetime | None:
    """Parse an ISO format string to datetime, or None."""
    return datetime.fromisoformat(s) if s else None


def _row_to_agent(row: aiosqlite.Row) -> Agent:
    """Convert a database row to an Agent dataclass."""
    return Agent(
        id=row[0],
        name=row[1],
        external_user_id=row[2],
        public_key=row[3],
        key_fingerprint=row[4],
        platform=row[5],
        status=row[6],
        enrolled_at=_str_to_dt(row[7]),
        last_auth_at=_str_to_dt(row[8]),
        key_expires_at=_str_to_dt(row[9]),
        created_at=_str_to_dt(row[10]) or datetime.utcnow(),
    )


class SQLiteStorage:
    """Async SQLite storage adapter implementing the StorageAdapter protocol."""

    def __init__(self, db_path: str = ":memory:") -> None:
        self._db_path = db_path
        self._db: aiosqlite.Connection | None = None

    async def _get_db(self) -> aiosqlite.Connection:
        if self._db is None:
            self._db = await aiosqlite.connect(self._db_path)
            await self._db.executescript(_CREATE_TABLES)
        return self._db

    # -- Agents --

    async def create_agent(self, agent: Agent) -> Agent:
        db = await self._get_db()
        await db.execute(
            """
            INSERT INTO sigil_agents
                (id, name, external_user_id, public_key, key_fingerprint,
                 platform, status, enrolled_at, last_auth_at, key_expires_at, created_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                agent.id,
                agent.name,
                agent.external_user_id,
                agent.public_key,
                agent.key_fingerprint,
                agent.platform,
                agent.status,
                _dt_to_str(agent.enrolled_at),
                _dt_to_str(agent.last_auth_at),
                _dt_to_str(agent.key_expires_at),
                _dt_to_str(agent.created_at),
            ),
        )
        await db.commit()
        return agent

    async def get_agent(self, agent_id: str) -> Agent | None:
        db = await self._get_db()
        cursor = await db.execute(
            "SELECT * FROM sigil_agents WHERE id = ?",
            (agent_id,),
        )
        row = await cursor.fetchone()
        return _row_to_agent(row) if row else None

    async def update_agent_status(
        self, agent_id: str, status: str, **fields: object
    ) -> None:
        db = await self._get_db()

        set_clauses = ["status = ?"]
        params: list[object] = [status]

        field_mapping = {
            "public_key": "public_key",
            "key_fingerprint": "key_fingerprint",
            "platform": "platform",
            "enrolled_at": "enrolled_at",
            "last_auth_at": "last_auth_at",
            "key_expires_at": "key_expires_at",
        }

        for field_name, column in field_mapping.items():
            if field_name in fields:
                value = fields[field_name]
                if isinstance(value, datetime):
                    value = _dt_to_str(value)
                set_clauses.append(f"{column} = ?")
                params.append(value)

        params.append(agent_id)
        sql = f"UPDATE sigil_agents SET {', '.join(set_clauses)} WHERE id = ?"
        await db.execute(sql, params)
        await db.commit()

    async def list_agents_by_user(self, user_id: str) -> list[Agent]:
        db = await self._get_db()
        cursor = await db.execute(
            "SELECT * FROM sigil_agents WHERE external_user_id = ?",
            (user_id,),
        )
        rows = await cursor.fetchall()
        return [_row_to_agent(row) for row in rows]

    # -- Enrollment Tokens --

    async def create_enrollment_token(self, token: EnrollmentToken) -> None:
        db = await self._get_db()
        await db.execute(
            """
            INSERT INTO sigil_enrollment_tokens (token_hash, agent_id, expires_at, used)
            VALUES (?, ?, ?, ?)
            """,
            (
                token.token_hash,
                token.agent_id,
                _dt_to_str(token.expires_at),
                1 if token.used else 0,
            ),
        )
        await db.commit()

    async def validate_enrollment_token(
        self, token_hash: str
    ) -> EnrollmentToken | None:
        db = await self._get_db()
        cursor = await db.execute(
            """
            SELECT token_hash, agent_id, expires_at, used
            FROM sigil_enrollment_tokens
            WHERE token_hash = ? AND used = 0
            """,
            (token_hash,),
        )
        row = await cursor.fetchone()
        if not row:
            return None

        token = EnrollmentToken(
            token_hash=row[0],
            agent_id=row[1],
            expires_at=_str_to_dt(row[2]) or datetime.utcnow(),
            used=bool(row[3]),
        )

        # Check expiration
        if token.expires_at < datetime.utcnow():
            return None

        return token

    async def burn_enrollment_token(self, token_hash: str) -> None:
        db = await self._get_db()
        await db.execute(
            "UPDATE sigil_enrollment_tokens SET used = 1 WHERE token_hash = ?",
            (token_hash,),
        )
        await db.commit()

    # -- Challenges --

    async def create_challenge(self, challenge: Challenge) -> None:
        db = await self._get_db()
        await db.execute(
            """
            INSERT INTO sigil_challenges (challenge, agent_id, expires_at, used)
            VALUES (?, ?, ?, ?)
            """,
            (
                challenge.challenge,
                challenge.agent_id,
                _dt_to_str(challenge.expires_at),
                1 if challenge.used else 0,
            ),
        )
        await db.commit()

    async def validate_challenge(self, challenge: str) -> Challenge | None:
        db = await self._get_db()
        cursor = await db.execute(
            """
            SELECT challenge, agent_id, expires_at, used
            FROM sigil_challenges
            WHERE challenge = ? AND used = 0
            """,
            (challenge,),
        )
        row = await cursor.fetchone()
        if not row:
            return None

        ch = Challenge(
            challenge=row[0],
            agent_id=row[1],
            expires_at=_str_to_dt(row[2]) or datetime.utcnow(),
            used=bool(row[3]),
        )

        if ch.expires_at < datetime.utcnow():
            return None

        return ch

    async def burn_challenge(self, challenge: str) -> None:
        db = await self._get_db()
        await db.execute(
            "UPDATE sigil_challenges SET used = 1 WHERE challenge = ?",
            (challenge,),
        )
        await db.commit()

    # -- Lifecycle --

    async def cleanup(self) -> None:
        if self._db:
            await self._db.close()
            self._db = None
